### PR TITLE
[CL Fees]: Add accumulator to lp positions

### PR DESCRIPTION
**Pull request by mattverse**
Closes: #[3926](https://github.com/osmosis-labs/osmosis/issues/3926)

## What is the purpose of the change

This issue focuses on implementing and testing the initializeFeeAccumulatorPosition and updateFeeAccumulatorPosition method and its dependency methods.

The methods are to be called in createPosition and updatePosition.

Additional minor method name change from `isInitialPosition` -> `isInitialPositionForPool` and `initializeInitialPosition` -> `initializeInitialPositionForPool`. 

## Brief Changelog

Add accumulators to lp positions (createPosition, updatePosition)

## Testing and Verifying
Unit tests has been added. 

The following two unit tests could not be added since this PR does not include the scope of setting feeGrowthOutside for each tick and these two methods have dependencies on such feature
- `updatePosition`
- `updateFeeAccumulatorPosition`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/concentrated-liquidity/lp.go [hunk 1]
```diff
@@ -187,6 +192,10 @@ func (k Keeper) updatePosition(ctx sdk.Context, poolId uint64, owner sdk.AccAddr
 		return sdk.Int{}, sdk.Int{}, err
 	}
 
+	if err := k.initOrUpdateFeeAccumulatorPosition(ctx, poolId, owner, liquidityDelta, lowerTick, upperTick); err != nil {
```
Comment (p0mvn) [inline x/concentrated-liquidity/lp.go:195]: I think we should keep this function to only be concerned with updates since it is called from both `createPosition` and `withdrawPosition`.

The "init" semantics here also feels redundant given that we initialize a position in `createPool` specifically anyways

File: x/concentrated-liquidity/lp.go [hunk 2]
```diff
@@ -47,19 +47,24 @@ func (k Keeper) createPosition(ctx sdk.Context, poolId uint64, owner sdk.AccAddr
...
-	}
 
-	// Calculate the amount of liquidity that will be added to the pool by creating this position.
-	liquidityDelta := math.GetLiquidityFromAmounts(pool.GetCurrentSqrtPrice(), sqrtPriceLowerTick, sqrtPriceUpperTick, amount0Desired, amount1Desired)
-	if liquidityDelta.IsZero() {
-		return sdk.Int{}, sdk.Int{}, sdk.Dec{}, errors.New("liquidityDelta calculated equals zero")
+		err := k.initializeInitialPositionForPool(cacheCtx, pool, amount0Desired, amount1Desired)
+		if err != nil {
+			return sdk.Int{}, sdk.Int{}, sdk.Dec{}, err
+		}
```
Comment (p0mvn) [inline x/concentrated-liquidity/lp.go:67]: Upon more testing, I found that to be correct, the call to `initializeFeeAccumulatorPosition` must be surrounded by the call to `hasPosition` instead of `isInitialPositionForPool`.

I updated this branch to reflect this: https://github.com/osmosis-labs/osmosis/pull/3893/files#diff-172bed5666a1e4a35522b031d46c6c7af3380aa208b753f336635ac046f1af1f

The reason is that `isInitialPosiitonForPool` returns true if this is the first position for pool in general. However, we want to call `initializeFeeAccumulatorPosition` only if this is the first time a given owner is adding liquidity for certain ticks. `hasPosition` does exactly that.

In other words, this is what I'm suggesting: https://github.com/osmosis-labs/osmosis/blob/5531b3f9e9a937eb1705d239c8ec18516ee6b4c4/x/concentrated-liquidity/lp.go#L65-L69

Please let me know if the suggestion makes sense

File: x/concentrated-liquidity/lp.go (hunk 1 above)
Comment (mattverse) [reply x/concentrated-liquidity/lp.go:199]: yeah, my first intention for this addition was to prevent erroring upon updating position wihtout having the fee accumulator initialized. But on the second thought, I agree with you that it is the correct behavior to error when fee accumulator position has not initialized

File: x/concentrated-liquidity/lp.go (hunk 2 above)
Comment (mattverse) [reply x/concentrated-liquidity/lp.go]: passing this for now with agreement that we will becoming back to lp.go changes in the future

File: x/concentrated-liquidity/lp.go [hunk 3]
```diff
@@ -46,11 +46,12 @@ func (k Keeper) createPosition(ctx sdk.Context, poolId uint64, owner sdk.AccAddr
 	cacheCtx, writeCacheCtx := ctx.CacheContext()
 	initialSqrtPrice := pool.GetCurrentSqrtPrice()
 	initialTick := pool.GetCurrentTick()
+	isInitialPositionForPool := k.isInitialPositionForPool(initialSqrtPrice, initialTick)
 
 	// If the current square root price and current tick are zero, then this is the first position to be created for this pool.
 	// In this case, we calculate the square root price and current tick based on the inputs of this position.
-	if k.isInitialPosition(initialSqrtPrice, initialTick) {
-		err := k.initializeInitialPosition(cacheCtx, pool, amount0Desired, amount1Desired)
+	if isInitialPositionForPool {
```
[the remark this answers is not in the document]
Comment (mattverse) [reply x/concentrated-liquidity/lp.go:53]: Removed all lp.go related logic

File: x/concentrated-liquidity/fees_test.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+				// position should have been properly initialzied to liquidityDelta provided
+				s.Require().Equal(positionSize, defaultLiquidityDelta)
+			} else {
+				s.Require().Error(err)
+			}
+		})
+	}
+}
+
+func (s *KeeperTestSuite) TestGetFeeGrowthOutside() {
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/concentrated-liquidity/fees_test.go:67]: Discussed offline - going to address separately

**Review by p0mvn — APPROVED** [review]
LGTM once linter passes

**Review by p0mvn — APPROVED** [review]
